### ISSUE TITLE
make material numbering more explicit

### ISSUE THREAD

**Issue by xjrc** [LLNL/conduit#676]
[Multi-buffer material sets](https://llnl-conduit.readthedocs.io/en/latest/blueprint_mesh.html#multi-buffer-material-sets) suffer from an issue wherein the numbering/enumeration order for the material set is ill-defined. Consider the following `matset` definition and its corresponding index entry:

```
# Schema #
domain0:
  matsets:
    matset:
      topology: topo
      volume_fractions:
        b: [ ... ]
        a: [ ... ]

# Index #
domain0:
  matsets:
    matset:
      topology: topo
      materials:
        b: (null)
        a: (null)
      path: matsets/matset
```

Since the base schema has no well-defined ordering for its materials, its entry within the index inherits this ill-defined information. Ascribing some sort of ordering to this information is important as it plays an important role in facilitating uniform presentation of this information in consumer codes (e.g. material identifiers in VisIt, Silo, etc.). This should be fixed at some level so that the ordering of these materials can be more consistent.

**Comment by xjrc** [comment]
One possible solution is to change `material_map` to being a conditional entry for uni-buffer material sets and make it an optional entry for all material sets, with the well-defined default of using lexicographical ordering to define this entry if it doesn't exist.

To help illustrate, here are some examples:

#### Explicit Specification ####

```
matsets:
  matset:
    topology: topology
    material_map:
      a: 1
      b: 0
      c: 2
    volume_fractions:
      a: [a0, a1, 0]
      b: [b0, b1, b2]
      c: [0, 0, c2]
```

#### Implicit Specification ####

```
matsets:
  matset:
    topology: topology
    volume_fractions:
      a: [a0, a1, 0]
      b: [b0, b1, b2]
      c: [0, 0, c2]
#   material_map:
#     a: 0
#     b: 1
#     c: 2
```

**Comment by cyrush** [comment]
Ok, hitting this now and need to resolve for the VisIt Blueprint Plugin.

I like the idea of having the optional material map. If it is there, it can be used and propagated to the blueprint index matset entry (as well as solving the other issues related to converting between formats, including `to_silo`)

If it is not there, we will infer an implicit map based on the order passed.

For this, I can change the behavior of `to_silo` to use the new implicit map, instead of sorting. 

@xjrc -- Does that sound good?

**Comment by xjrc** [comment]
@cyrush: Yes, that sounds good. In order to guarantee uniformity between the index and the mesh order, I think we should also update `generate_index` to use this inferred map when populating the material index section. If that and `to_silo` are both updated, I think we'll have hit every interface that depends on this.

**Comment by cyrush** [comment]
resolved in #685
